book/src/input/how_to_define_custom_control_bindings.md

```diff
@@ -0,0 +1,172 @@
+* Inputting a `String` when retrieving input data is error-prone if you mistype it or change the name.
+* You will be able to add additional information to a custom type.
+
+## Defining a Custom Type
+
+Defining a custom type for the `InputBundle` is done by implementing the `BindingTypes` trait. This trait contains two types, an `Axis` type and an `Action` type. These types are usually defined as enums, but you can use whatever you want.
+
+```rust,edition2018,no_run,noplaypen
+use serde::{Serialize, Deserialize};
+use amethyst::input::{BindingTypes, Bindings};
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq, Serialize, Deserialize)]
+enum AxisBinding {
+  Horizontal,
+  Vertical,
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq, Serialize, Deserialize)]
+enum ActionBinding {
+  Shoot,
+}
+
+impl Display for AxisBinding {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl Display for ActionBinding {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+struct MovementBindingTypes;
+
+impl BindingTypes for MovementBindingTypes {
+  type Axis = AxisBinding;
+  type Action = ActionBinding;
+}
+```
+
+The `Axis` and `Action` type both need to derive all the traits listed above, the first five are used by Amethyst and the last two are for reading and writing to files correctly. They also need to implement `Display` if you want to add a bindings config file.
+
+For serializing and deserializing you need to add [serde](https://crates.io/crates/serde) to the dependencies like this (note that there might be a newer version available, that is also compatible):
+
+```rust,ignore
+  serde = { version="1.0.92", features = ["derive"] }
+```
+
+If you want to add additional information you can add it to the enum or change the `Axis` and `Action` types to a struct.
+
+```rust,edition2018,no_run,noplaypen
+#[derive(Clone, Debug, Hash, PartialEq, Eq, Serialize, Deserialize)]
+enum AxisBinding {
+  Horizontal(usize),
+  Vertical(usize),
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq, Serialize, Deserialize)]
+enum ActionBinding {
+  Shoot(usize),
+}
+
+//..
+```
+
+We can now use this custom type in our `InputBundle` and create a RON config file for our bindings.
+
+The config file might look something like this:
+
+```ron,ignore
+(
+  axes: {
+    Vertical(0): Emulated(pos: Key(W), neg: Key(S)),
+    Horizontal(0): Emulated(pos: Key(D), neg: Key(A)),
+    Vertical(1): Emulated(pos: Key(Up), neg: Key(Down)),
+    Horizontal(1): Emulated(pos: Key(Right), neg: Key(Left)),
+  },
+  actions: {
+    Shoot(0): [[Key(Space)]],
+    Shoot(1): [[Key(Return)]],
+  },
+)
+```
+
+Here the number after the binding type could be the ID of the player, but you can supply any other data as long as it derives the right traits.
+
+With the config file we can create an `InputBundle` like in the previous section.
+
+```rust,edition2018,no_run,noplaypen
```

Comment:
`rust,ignore` here would let this pass the mdbook check, next snippet could also be ignored